### PR TITLE
[REG-1011] Create a draft bot-validation strategy

### PR DESCRIPTION
**Pull request by abeizer**
Requires GameMachine PR: https://github.com/Regression-Games/RegressionGames/pull/320

Example of how validations appear for "automated test" style bots. This example shows the menu bot clicking through buttons and making assertions about the state after performing each action:
https://www.loom.com/share/7ff0288b57634a45b46fe6cbdfd3c5d3?sid=2d1c6c07-8747-4a8e-9452-4a722f6fe69d

Example of how validations appear for "dynamic" style bots. This example shows the co-op bot using validations to make decisions rather than test-style assertions:
https://www.loom.com/share/617b7fbce4024aee9729451ecfcc6f0d?sid=74d79f8f-c4bd-40af-a48b-f717af0aa959

### CODE REVIEW

File: menubot/index.js [hunk 1]
```diff
@@ -1,85 +1,102 @@
...
 
-export async function runTurn(playerId, tickInfo, mostRecentMatchInfo, actionQueue) {
-
-  // On each state, run through the validations and see if any failed or passed
-  rgValidator.checkValidations(tickInfo);
-
-  const t = tickInfo.tick;
-  const sceneName = tickInfo.sceneName;
+export async function configureBot(rgObject) {
+  rg = rgObject;
```
Comment (RG-nAmKcAz) [inline menubot/index.js:50]: i feel like this should be a no-op ?

why wouldn't we use the rgObject reference passed into runTurn

File: menubot/index.js [hunk 2]
```diff
@@ -88,74 +105,32 @@ export async function runTurn(playerId, tickInfo, mostRecentMatchInfo, actionQue
...
-  const buttons = RGBot.getEntitiesOfType(tickInfo, buttonName);
-  //console.log(`Found buttons: ${JSON.stringify(buttons)}`)
-  if (buttons && buttons.length > 0) {
-    const button = buttons[0];
-    if (button.interactable) {
-      //console.log(`Returning button: ${JSON.stringify(button)}`)
-      return button;
-    }
+async function getInteractableButton(buttonName) {
+  const button = await rg.findEntity(buttonName);
```
Comment (RG-nAmKcAz) [inline menubot/index.js:131]: I feel like this should be named `waitForEntity` and have some configurable timeout value.

**Review by RG-nAmKcAz — COMMENTED** [review]
After seeing this, it may clarify some of my comments around findEntity in the Other pr

File: menubot/index.js (hunk 2 above)
Comment (abeizer) [reply menubot/index.js:131]: In non-test use cases (where you'd use runTurn) it doesn't actually wait for that entity to exist, it simply evaluates the state at that tick like the current implementation of the menubot does

File: menubot/index.js (hunk 1 above)
Comment (abeizer) [reply menubot/index.js]: Good call, updated

File: abilitybot/configureBotExample.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+export async function configureBot(rgObject) {
+    rg = rgObject;
+    rg.automatedTestMode = true;
+
+    // validate that we're in the game
+    await rg.waitForScene("BossRoom");
+
+    // find the closest human player and use a heal ability on them
+    let target = await rg.findNearestEntity("HumanPlayer");
```
Comment (vontell) [inline abilitybot/configureBotExample.js:42]: Doesn't findNearestEntity already fail if it doesn't exist? So the second line may be redundant?

File: abilitybot/configureBotExample.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    rg.performAction("PerformSkill", args)
+    await rg.entityHasAttribute(target, "health", 10);
+
+    rg.performAction("PerformSkill", args)
+    await rg.entityHasAttribute(target, "health", 5);
+
+    rg.performAction("PerformSkill", args)
+    await rg.entityDoesNotExist(target);
+
+    rg.complete();
```
Comment (vontell) [inline abilitybot/configureBotExample.js:90]: By default we could always apply this after the end of configureBot if there is no runTurn

**Review by vontell — COMMENTED** [review]
Not too many comments on this, I think I may have more after the comments in the other PR have responses though

File: abilitybot/configureBotExample.js (hunk 3 above)
Comment (abeizer) [reply abilitybot/configureBotExample.js:42]: Today we don't stop the bot when a validation fails. For test bots I'd love this behavior, but dynamic bots shouldn't stop if one of their evaluations fail.

**Review by vontell — APPROVED** [review]

**Pull request merged**